packages/firestore/src/local/memory_persistence.ts

```diff
@@ -119,3 +154,154 @@ export class MemoryPersistence implements Persistence {
+  }
+
+  getTargetCount(txn: PersistenceTransaction): PersistencePromise<number> {
+    return this.persistence.getQueryCache().getTargetCount(txn);
+  }
+
+  forEachOrphanedDocumentSequenceNumber(
+    txn: PersistenceTransaction,
+    f: (sequenceNumber: ListenSequenceNumber) => void
+  ): PersistencePromise<void> {
+    this.orphanedSequenceNumbers.forEach((_, sequenceNumber) =>
+      f(sequenceNumber)
+    );
+    return PersistencePromise.resolve();
+  }
+
+  setInMemoryPins(inMemoryPins: ReferenceSet): void {
+    this.additionalReferences = inMemoryPins;
+  }
+
+  removeTargets(
+    txn: PersistenceTransaction,
+    upperBound: ListenSequenceNumber,
+    activeTargetIds: ActiveTargets
+  ): PersistencePromise<number> {
+    return this.persistence
+      .getQueryCache()
+      .removeTargets(txn, upperBound, activeTargetIds);
+  }
+
+  removeOrphanedDocuments(
+    txn: PersistenceTransaction,
+    upperBound: ListenSequenceNumber
+  ): PersistencePromise<number> {
+    let count = 0;
+    const cache = this.persistence.getRemoteDocumentCache();
+    const p = cache.forEachDocumentKey(txn, key => {
+      return this.isPinned(txn, key, upperBound).next(isPinned => {
+        if (isPinned) {
+          return PersistencePromise.resolve();
+        } else {
+          count++;
+          return cache.removeEntry(txn, key);
+        }
+      });
+    });
+    return p.next(() => count);
+  }
+
+  removeMutationReference(
+    txn: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<void> {
+    this.orphanedSequenceNumbers.set(key, txn.currentSequenceNumber);
+    return PersistencePromise.resolve();
+  }
+
+  removeTarget(
+    txn: PersistenceTransaction,
+    queryData: QueryData
+  ): PersistencePromise<void> {
+    const updated = queryData.copy({
+      sequenceNumber: txn.currentSequenceNumber
+    });
+    return this.persistence.getQueryCache().updateQueryData(txn, updated);
+  }
+
+  addReference(
+    txn: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<void> {
+    this.orphanedSequenceNumbers.set(key, txn.currentSequenceNumber);
+    return PersistencePromise.resolve();
+  }
+
+  removeReference(
+    txn: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<void> {
+    this.orphanedSequenceNumbers.set(key, txn.currentSequenceNumber);
+    return PersistencePromise.resolve();
+  }
+
+  updateLimboDocument(
+    txn: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<void> {
+    this.orphanedSequenceNumbers.set(key, txn.currentSequenceNumber);
+    return PersistencePromise.resolve();
+  }
+
+  private isPinned(
+    txn: PersistenceTransaction,
+    key: DocumentKey,
+    upperBound: ListenSequenceNumber
+  ): PersistencePromise<boolean> {
+    return or([
+      () => this.persistence.mutationQueuesContainKey(txn, key),
+      () => this.additionalReferences!.containsKey(txn, key),
```

Comment:
Is `additionalReferences` guaranteed to be set here? We could also initialize it with an empty set.

Reply:
It is required to be set. It originally could be set multiple times in spec tests, but that might actually be gone now. I'm going to leave this consistent with other ports but file a bug to investigate if this could be an argument to the constructor: b/116248679